src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs

```diff
@@ -424,56 +409,56 @@ internal SymbolicRegexNode<TNewSet> Transform<TNewSet>(SymbolicRegexNode<TSet> s
+                    Debug.Assert(node._left is not null);
+                    return builder.CreateLoop(Transform(node._left, builder, setTransformer), node.IsLazy, node._lower, node._upper);
 
                 case SymbolicRegexNodeKind.Or:
-                    Debug.Assert(sr._alts is not null);
-                    return builder.Or(sr._alts.Transform(builder, setTransformer));
+                    Debug.Assert(node._alts is not null);
+                    return builder.Or(node._alts.Transform(builder, setTransformer));
 
                 case SymbolicRegexNodeKind.OrderedOr:
-                    Debug.Assert(sr._left is not null && sr._right is not null);
-                    return builder.OrderedOr(Transform(sr._left, builder, setTransformer), Transform(sr._right, builder, setTransformer));
+                    Debug.Assert(node._left is not null && node._right is not null);
+                    return builder.OrderedOr(Transform(node._left, builder, setTransformer), Transform(node._right, builder, setTransformer));
 
                 case SymbolicRegexNodeKind.And:
-                    Debug.Assert(sr._alts is not null);
-                    return builder.And(sr._alts.Transform(builder, setTransformer));
+                    Debug.Assert(node._alts is not null);
+                    return builder.And(node._alts.Transform(builder, setTransformer));
 
                 case SymbolicRegexNodeKind.CaptureStart:
-                    return builder.CreateCaptureStart(sr._lower);
+                    return builder.CreateCaptureStart(node._lower);
 
                 case SymbolicRegexNodeKind.CaptureEnd:
-                    return builder.CreateCaptureEnd(sr._lower);
+                    return builder.CreateCaptureEnd(node._lower);
 
                 case SymbolicRegexNodeKind.Concat:
                     {
-                        List<SymbolicRegexNode<TSet>> sr_elems = sr.ToList();
-                        SymbolicRegexNode<TNewSet>[] sr_elems_trasformed = new SymbolicRegexNode<TNewSet>[sr_elems.Count];
-                        for (int i = 0; i < sr_elems.Count; i++)
+                        List<SymbolicRegexNode<TSet>> concatElements = node.ToList();
+                        SymbolicRegexNode<TNewSet>[] reverseTransformed = new SymbolicRegexNode<TNewSet>[concatElements.Count];
+                        for (int i = 0; i < reverseTransformed.Length; i++)
                         {
-                            sr_elems_trasformed[i] = Transform(sr_elems[i], builder, setTransformer);
+                            reverseTransformed[i] = Transform(concatElements[^(i + 1)], builder, setTransformer);
                         }
-                        return builder.CreateConcat(sr_elems_trasformed, false);
+                        return builder.CreateConcatAlreadyReversed(reverseTransformed, tryCreateFixedLengthMarker: false);
```

Comment:
I reversed the order of filling the array here, such that we could then get rid of the CreateConcat overload which itself iterates through the array in reverse order and delegates to this one.